icicle/backend/cpu/src/hash/cpu_poseidon2.cpp

```diff
@@ -145,341 +142,437 @@ namespace icicle {
-        partial_matrix_diagonal_m1_mul_by_vector(tmp_fields, partial_matrix_diagonal_m1, tmp_fields);
+      S prepared_element;
+#pragma unroll
+      for (int element_idx_in_hash = 0; element_idx_in_hash < T; element_idx_in_hash++) {
+        if (use_domain_tag) {
+          if (element_idx_in_hash == 0) {
+            prepared_element = domain_tag;
+          } else {
+            prepared_element = input[element_idx_in_hash - 1];
+          }
+        } else {
+          prepared_element = input[element_idx_in_hash];
+        }
+        states[element_idx_in_hash] = prepared_element;
       }
+    } // prepare_poseidon2_states(const S* input, S* states, bool use_domain_tag, S domain_tag)
+
+    template <int T>
+    void prepare_poseidon2_sponge_states(
+      const S* input,
+      const bool is_padding_needed,
+      const S* padding,
+      const int padding_size_in_scalars,
+      S* states,
+      const bool use_domain_tag,
+      const S domain_tag,
+      const bool is_first_hasher,
+      const bool is_last_hasher) const
+    {
+      if (is_first_hasher) {
+        if (use_domain_tag) {
+          states[0] = domain_tag;
+        } else {
+          states[0] = input[0];
+          input += 1; // Increment the pointer. Now rest of the input is of T-1 granularity.
+        }
+      }
+      // For middle and last hashers states[0] stays the same as the output of the prev hasher. So, no need to take care
+      // of that.
+#pragma unroll
+      for (int states_idx = 1; states_idx < T; states_idx++) {
+        if (is_last_hasher && is_padding_needed) {
+          int nof_valid_inputs_in_hashers = (T - 1) - padding_size_in_scalars;
+          if (states_idx < 1 + nof_valid_inputs_in_hashers)
+            states[states_idx] = states[states_idx] + input[states_idx - 1];
+          else
+            states[states_idx] = states[states_idx] + padding[states_idx - (1 + nof_valid_inputs_in_hashers)];
+        } else { // Not last hasher or no padding.
+          states[states_idx] = states[states_idx] + input[states_idx - 1];
+        }
+      }
+    } // prepare_poseidon2_sponge_states(const S* input, S* states, unsigned int batch_size, bool use_domain_tag, S
+      // domain_tag)
 
-      // Bottom full rounds.
-      full_rounds(nof_bottom_full_rounds, tmp_fields, rounds_constants);
-
-      memcpy(hasher_output, (std::byte*)(tmp_fields), T * sizeof(S));
-      // memcpy(output, (std::byte*)(&tmp_fields[1]), sizeof(S));
-
-      return eIcicleError::SUCCESS;
-    } // eIcicleError hash_single(const std::byte* input, std::byte* output) const
-
-    // This function performs a partial_matrix_diagonal_m1 matrix by vector multiplication.
-    // Note that in order to increase the performance the partial matrix diagonal values
-    // are actually partial matrix diagonal values minus 1.
-    void partial_matrix_diagonal_m1_mul_by_vector(const S* vec_in, const S* matrix_in, S* result) const
+    S sbox_el(S element, const int alpha) const
```

Comment:
name unclear, sbox_element?
```suggestion
    eIcicleError sbox_element(S element, const int alpha, S to_return) const
```

Reply:
Done.